libraries/AP_HAL/utility/BetterStream.h

```diff
@@ -36,11 +36,11 @@ class AP_HAL::BetterStream {
 
     virtual size_t write(uint8_t) = 0;
     virtual size_t write(const uint8_t *buffer, size_t size) = 0;
-    size_t write(const char *str);
+    virtual size_t write(const char *str);
 
     virtual uint32_t available() = 0;
 
-    int16_t read();  // old function; prefer calling the boolean method
```

Comment:
The comment being removed is still valid.

AFAICS you've just duplicated this function onto `AP_HAL::UARTDriver` - adding to code size without benefit.

This method could just be removed from here and moved into UARTDriver; the two changes required to make this work are trivial and make the affected code look a lot better:

```
pbarker@fx:~/rc/ardupilot((HEAD detached at tridge/pr-uCenter-over-DroneCAN))$ git diff
diff --git a/libraries/AP_HAL/UARTDriver.h b/libraries/AP_HAL/UARTDriver.h
index 9e4d23b5dea..44b82029217 100644
--- a/libraries/AP_HAL/UARTDriver.h
+++ b/libraries/AP_HAL/UARTDriver.h
@@ -51,7 +51,7 @@ public:
     /*
       single and multi-byte read methods
      */
-    int16_t read(void) override;
+    int16_t read(void);
     bool read(uint8_t &b) override WARN_IF_UNUSED;
     ssize_t read(uint8_t *buffer, size_t count) override;
     
diff --git a/libraries/AP_HAL/utility/BetterStream.cpp b/libraries/AP_HAL/utility/BetterStream.cpp
index d30908146de..ba966262234 100644
--- a/libraries/AP_HAL/utility/BetterStream.cpp
+++ b/libraries/AP_HAL/utility/BetterStream.cpp
@@ -20,24 +20,14 @@ size_t AP_HAL::BetterStream::write(const char *str)
     return write((const uint8_t *)str, strlen(str));
 }
 
-int16_t AP_HAL::BetterStream::read()
-{
-    uint8_t b;
-    if (!read(b)) {
-        return -1;
-    }
-    return b;
-}
-
 ssize_t AP_HAL::BetterStream::read(uint8_t *buffer, size_t count)
 {
     size_t offset = 0;
     while (count--) {
-        const int16_t x = read();
-        if (x == -1) {
+        if (!read(buffer[offset])) {
             return offset;
         }
-        buffer[offset++] = (uint8_t)x;
+        offset++;
     }
     return offset;
 }
diff --git a/libraries/AP_HAL/utility/BetterStream.h b/libraries/AP_HAL/utility/BetterStream.h
index 694ac908671..2e046b0a5b1 100644
--- a/libraries/AP_HAL/utility/BetterStream.h
+++ b/libraries/AP_HAL/utility/BetterStream.h
@@ -40,7 +40,6 @@ public:
 
     virtual uint32_t available() = 0;
 
-    virtual int16_t read(void);
     virtual bool read(uint8_t &b) WARN_IF_UNUSED = 0;
 
     // no base-class implementation to force descendants to
diff --git a/libraries/GCS_MAVLink/GCS_serial_control.cpp b/libraries/GCS_MAVLink/GCS_serial_control.cpp
index 133297dbcbb..5d761a27b61 100644
--- a/libraries/GCS_MAVLink/GCS_serial_control.cpp
+++ b/libraries/GCS_MAVLink/GCS_serial_control.cpp
@@ -181,7 +181,10 @@ more_data:
      }
 
     // read any reply data
-    packet.count = 0;
     memset(packet.data, 0, sizeof(packet.data));
-    while (available > 0) {
-        packet.data[packet.count++] = (uint8_t)stream->read();
-        available--;
-    }
+    packet.count = stream->read(&packet.data[0], available);
 
     // and send the reply
     _mav_finalize_message_chan_send(chan, 

pbarker@fx:~/rc/ardupilot((HEAD detached at tridge/pr-uCenter-over-DroneCAN))$ 
```

(only tested as far as "it compiled once, on my machine")